File: goban.base/src/com/computernerd1101/goban/internal/AtomicArray52.java
package com.computernerd1101.goban.internal;

import org.jetbrains.annotations.Nullable;

import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;

@SuppressWarnings({"unchecked", "rawtypes", "unused"})
public class AtomicArray52<E> {

    public static final AtomicIntegerFieldUpdater<? super AtomicArray52<?>> SIZE =
            AtomicIntegerFieldUpdater.newUpdater(AtomicArray52.class, "size");

    private static final AtomicReferenceFieldUpdater[] UPDATE = new AtomicReferenceFieldUpdater[52];

    static {
        char[] buffer = new char[3];
        buffer[0] = 'e';
        for(int index = 0; index < 52; index++) {
            int n;
            if (index < 10) {
                buffer[1] = (char)('0' + index);
                n = 2;
            } else {
                buffer[1] = (char)('0' + index / 10);
                buffer[2] = (char)('0' + index % 10);
                n = 3;
            }
            UPDATE[index] = AtomicReferenceFieldUpdater.newUpdater(
                    AtomicArray52.class, Object.class, new String(buffer, 0, n).intern());
        }
    }

    @SuppressWarnings("unchecked")
    public static <E> AtomicReferenceFieldUpdater<AtomicArray52<E>, E> update(int index) {
        return (AtomicReferenceFieldUpdater<AtomicArray52<E>, E>)UPDATE[index];
    }

    @Nullable
    public final E get(int index) { return (E)UPDATE[index].get(this); }
    public final void set(int index, @Nullable E e) { UPDATE[index].set(this, e); }

    public volatile int size = 0;

    @Nullable public volatile E e0 = null;
    @Nullable public volatile E e1 = null;
    @Nullable public volatile E e2 = null;
    @Nullable public volatile E e3 = null;
    @Nullable public volatile E e4 = null;
    @Nullable public volatile E e5 = null;
    @Nullable public volatile E e6 = null;
    @Nullable public volatile E e7 = null;
    @Nullable public volatile E e8 = null;
    @Nullable public volatile E e9 = null;
    @Nullable public volatile E e10 = null;
    @Nullable public volatile E e11 = null;
    @Nullable public volatile E e12 = null;
    @Nullable public volatile E e13 = null;
    @Nullable public volatile E e14 = null;
    @Nullable public volatile E e15 = null;
    @Nullable public volatile E e16 = null;
    @Nullable public volatile E e17 = null;
    @Nullable public volatile E e18 = null;
    @Nullable public volatile E e19 = null;
    @Nullable public volatile E e20 = null;
    @Nullable public volatile E e21 = null;
    @Nullable public volatile E e22 = null;
    @Nullable public volatile E e23 = null;
    @Nullable public volatile E e24 = null;
    @Nullable public volatile E e25 = null;
    @Nullable public volatile E e26 = null;
    @Nullable public volatile E e27 = null;
    @Nullable public volatile E e28 = null;
    @Nullable public volatile E e29 = null;
    @Nullable public volatile E e30 = null;
    @Nullable public volatile E e31 = null;
    @Nullable public volatile E e32 = null;
    @Nullable public volatile E e33 = null;
    @Nullable public volatile E e34 = null;
    @Nullable public volatile E e35 = null;
    @Nullable public volatile E e36 = null;
    @Nullable public volatile E e37 = null;
    @Nullable public volatile E e38 = null;
    @Nullable public volatile E e39 = null;
    @Nullable public volatile E e40 = null;
    @Nullable public volatile E e41 = null;
    @Nullable public volatile E e42 = null;
    @Nullable public volatile E e43 = null;
    @Nullable public volatile E e44 = null;
    @Nullable public volatile E e45 = null;
    @Nullable public volatile E e46 = null;
    @Nullable public volatile E e47 = null;
    @Nullable public volatile E e48 = null;
    @Nullable public volatile E e49 = null;
    @Nullable public volatile E e50 = null;
    @Nullable public volatile E e51 = null;
    @Nullable public volatile E e52 = null;

}

File: goban.base/src/com/computernerd1101/goban/internal/AtomicArray52.kt
package com.computernerd1101.goban.internal

import java.util.concurrent.atomic.AtomicReferenceFieldUpdater

@Suppress("unused")
internal open class AtomicArray52<E: Any> {

    companion object {

        @JvmField val SIZE = atomicIntUpdater<AtomicArray52<*>>("size")

        @JvmField
        val UPDATE: Array<AtomicReferenceFieldUpdater<AtomicArray52<*>, *>> = CharArray(3).let { buffer ->
            buffer[0] = 'e'
            Array(52) { index ->
                val n = if (index < 10) {
                    buffer[1] = '0' + index
                    2
                } else {
                    buffer[1] = '0' + index / 10
                    buffer[2] = '0' + index % 10
                    3
                }
                atomicUpdater<AtomicArray52<*>, Any?>(buffer.concatToString(0, n))
            }
        }

        @Suppress("UNCHECKED_CAST")
        fun <E: Any> update(index: Int): AtomicReferenceFieldUpdater<AtomicArray52<E>, E?> =
            UPDATE[index] as AtomicReferenceFieldUpdater<AtomicArray52<E>, E?>

    }

    operator fun get(index: Int): E?    = update<E>(index)[this]
    operator fun set(index: Int, e: E?) { update<E>(index)[this] = e }

    @Volatile @JvmField var size: Int = 0

    @Volatile @JvmField var e0:  E? = null
    @Volatile @JvmField var e1:  E? = null
    @Volatile @JvmField var e2:  E? = null
    @Volatile @JvmField var e3:  E? = null
    @Volatile @JvmField var e4:  E? = null
    @Volatile @JvmField var e5:  E? = null
    @Volatile @JvmField var e6:  E? = null
    @Volatile @JvmField var e7:  E? = null
    @Volatile @JvmField var e8:  E? = null
    @Volatile @JvmField var e9:  E? = null
    @Volatile @JvmField var e10: E? = null
    @Volatile @JvmField var e11: E? = null
    @Volatile @JvmField var e12: E? = null
    @Volatile @JvmField var e13: E? = null
    @Volatile @JvmField var e14: E? = null
    @Volatile @JvmField var e15: E? = null
    @Volatile @JvmField var e16: E? = null
    @Volatile @JvmField var e17: E? = null
    @Volatile @JvmField var e18: E? = null
    @Volatile @JvmField var e19: E? = null
    @Volatile @JvmField var e20: E? = null
    @Volatile @JvmField var e21: E? = null
    @Volatile @JvmField var e22: E? = null
    @Volatile @JvmField var e23: E? = null
    @Volatile @JvmField var e24: E? = null
    @Volatile @JvmField var e25: E? = null
    @Volatile @JvmField var e26: E? = null
    @Volatile @JvmField var e27: E? = null
    @Volatile @JvmField var e28: E? = null
    @Volatile @JvmField var e29: E? = null
    @Volatile @JvmField var e30: E? = null
    @Volatile @JvmField var e31: E? = null
    @Volatile @JvmField var e32: E? = null
    @Volatile @JvmField var e33: E? = null
    @Volatile @JvmField var e34: E? = null
    @Volatile @JvmField var e35: E? = null
    @Volatile @JvmField var e36: E? = null
    @Volatile @JvmField var e37: E? = null
    @Volatile @JvmField var e38: E? = null
    @Volatile @JvmField var e39: E? = null
    @Volatile @JvmField var e40: E? = null
    @Volatile @JvmField var e41: E? = null
    @Volatile @JvmField var e42: E? = null
    @Volatile @JvmField var e43: E? = null
    @Volatile @JvmField var e44: E? = null
    @Volatile @JvmField var e45: E? = null
    @Volatile @JvmField var e46: E? = null
    @Volatile @JvmField var e47: E? = null
    @Volatile @JvmField var e48: E? = null
    @Volatile @JvmField var e49: E? = null
    @Volatile @JvmField var e50: E? = null
    @Volatile @JvmField var e51: E? = null

}